File: Python Solutions/School Round 2016/Round 2/Problem 3 Word Chain Solution 1.py
from string import ascii_uppercase

def diffrences(w1,w2):
    for l in range(len(w1)):
        if w1[l] != w2[l]:
            yield l

def one_diffrent(w1,w2):
    return len(list(diffrences(w1,w2))) == 1

class WordChain:
    def __init__(self,w1,w2):
        self.w1,self.w2 = w1,w2

    def __hash__(self):
        index = list(diffrences(self.w1,self.w2))[0]
        if ascii_uppercase.index(self.w1[index])>ascii_uppercase.index(self.w2[index]):
            return hash((self.w2,self.w1))
        else:
            return hash((self.w1,self.w2))

    def __repr__(self):
        return self.w1 + ' '+self.w2

    def __eq__(self, other):
        return self.w1 == other or self.w2 == other

    def other(self,word):
        if self.w1 == word:
            return self.w2
        else:
            return self.w1

def alphabet(words):
    words.sort()
    return words[0]

if __name__ == '__main__':
    connects = set()
    words = input('Input: ').split(' ')
    counts = [0 for _ in range(len(words))]
    for w, w1 in enumerate(words):
        for w2 in words[w+1:]:
            if one_diffrent(w1,w2):
                connects.add(WordChain(w1,w2))
    for w, w1 in enumerate(words):
        for connection in connects:
            if w1 == connection:
                counts[w]+=1
    key_words = [word for w,word in enumerate(words) if counts[w] == min(counts)]
    chains = [[alphabet(key_words)]]
    while len(chains[0]) < len(words):
        chains_copy = chains.copy()
        chains.clear()
        for chain in chains_copy:
            word = chain[-1]
            w = words.index(word)
            for connection in connects:
                if word == connection and connection.other(word) not in chain:
                    chains.append([*chain,connection.other(word)])
    print(' '.join(chains[0]))
'''
WALL TAIL TALL WALK BALL WAIL TALK
'''

File: C++_Solutions/School_Round_2016/Round_2/Problem_3_Word_Chain_Solution_1.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

string alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";

class Chain{
public:
  vector<string> chain;
  string last;

  Chain(const string &last){
    chain = vector<string>(0);
    chain.push_back(last);
    this->last = last;
  }

  Chain(const vector<string> &prev,const string &last){
    this->last = last;
    chain = prev;
    chain.push_back(last);
  }

  bool contains(const string &word){
    for(string w:chain){
      if(word == w) return true;
    }
    return false;
  }
};

bool is_equv(const string& one, const string &two){
  int errors = 0;
  for (size_t i = 0; i < one.size(); i++) {
    if(errors > 1) return false;
    if(one[i] != two[i]) errors += 1;
  }
  return errors == 1;
}


bool alpha(const string& word, const string& next){
  for (size_t i = 0; i < word.size(); i++) {
    if(alphabet.find(word[i])<alphabet.find(next[i])) return false;
    else if(alphabet.find(word[i])>alphabet.find(next[i])) return true;
  }
  return false;
}
bool better(const Chain& best, const Chain& next){
  for (size_t i = 0; i < best.chain.size(); i++) {
    if(best.chain[i] != next.chain[i]) return alpha(best.chain[i],next.chain[i]);
  }
}

int main(int argc, char const *argv[]) {
  vector<string> words = {"WALL","TAIL","TALL","WALK","BALL","WAIL","TALK"};
  vector<Chain*> chains(0);
  for(auto word:words) chains.push_back(new Chain(word));
  for (size_t i = 0; i < words.size()-1; i++) {
    vector<Chain*> chains_copy = chains;
    chains.clear();
    for(Chain* _chain: chains_copy){
      Chain& chain = *_chain;
      for(string word:words){
        if(!chain.contains(word)){
          if(is_equv(word,chain.last)) chains.push_back(new Chain(chain.chain,word));
        }
      }
    }
  }
  Chain& best = *chains[0];
  for (size_t i = 0; i < chains.size(); i++) {
    if(better(best,*chains[i])) best = *chains[i];
  }
  if(chains.size() == 0) cout<<"IMPOSSIBLE";
  else{
    for(auto word:best.chain) cout<<word<<' ';
    cout<<endl;
  }
  return 0;
}
